src/components/interactive/Button.jsx

```diff
@@ -0,0 +1,78 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import baseElement from '../../util/baseElement';
+
+const Button = ({
+  theme,
+  isLoading,
+  iconPlacement,
+  Icon,
+  Link,
+  disabled,
+  children,
+  ...rest
+}) => {
+  const Element = baseElement({ href: rest.href, as: Link });
+  return (
+    <Element
+      {...rest}
+      className={cx('fdsButton', {
+        'fdsButton--contained-blue': theme === 'contained--blue',
+        'fdsButton--contained-red': theme === 'contained--red',
+        'fdsButton--loading': isLoading,
+        'fdsButton--disabled': disabled,
+      })}
+      disabled={disabled && Element === 'button'}
+    >
+      <span className={cx({ 'fdsButton--hidden': isLoading })}>{children}</span>
+      {Icon && (
+        <div
+          className={cx('fdsButton-icon', {
+            'fdsButton-icon--left': iconPlacement === 'left',
+            'margin--left--half': iconPlacement === 'right',
+            'margin--right--half': iconPlacement === 'left',
+            'fdsButton--hidden': isLoading,
+          })}
+        >
+          <Icon size="xs" />
+        </div>
+      )}
+    </Element>
+  );
+};
+
+Button.defaultProps = {
+  theme: 'primary--blue',
```

Comment:
*tl;dr* - we should talk to design about this a bit more. We should resolve these questions:

- Supporting every color/type combination will provide a more flexible long-term solution. How does design feel about this? The tradeoff of not doing this means that we can never design a button that is any color other than blue or red (unless we refactor the API to what I am suggesting)
- When we go back to replace buttons in existing features, should we force all of them to blue or red?

----

I guess it would look more like...

```js
{
  type: PropTypes.oneOf('contained', 'outline', 'ghost'),
  color: PropTypes.oneOf('blue', 'red', ... ),
}
```

I have a few reasons for bringing this up:
- being able to independently declare type and color is scalable while maintaining a clear distinction between color and type
- the `theme` prop values introduce a new concept, which is a string that looks like a class but isn't actually a class
- the `theme` prop is behaving like two separate props in one, which is probably an indication it should be two props
- if we support some combinations and not others, this could get confusing to devs. If i see `primary--blue` and `ghost--red`, i might expect there would be a `ghost--blue`, but there's not. I'd have to look up docs every time I use a button to see what's valid.

I understand why you set it up this way though - there are limitations in color/type combinations that are dictated by the design.

On taking a closer look at designs, I suspect this won't scale. If we ever want a purple button for any reason, we're forcing developers to make a custom button, which eschews all of the resilience and best practices we build into the standard component.

Sometimes guard rails are nice, but if they're too restrictive, it encourages "off-roading" if a designer or engineer doesn't like where the component is telling them where to go.

Reply:
With all that said, our release habits mean that we can in fact make a breaking change to this later to support a different API if we need it. Let's discuss this first with design and make a call from there.